File: tests/test-terraria/src/main/java/pize/tests/terraria/entity/Player.java
package pize.tests.terraria.entity;

import pize.Pize;
import pize.graphics.util.batch.TextureBatch;
import pize.io.glfw.Key;
import pize.math.Maths;
import pize.math.vecmath.vector.Vec2f;
import pize.physic.BoundingBox2;
import pize.physic.Collider2f;
import pize.physic.RectBody;
import pize.tests.terraria.map.MapTile;
import pize.tests.terraria.map.WorldMap;
import pize.tests.terraria.world.World;

import java.util.ArrayList;
import java.util.List;

public class Player extends Entity{

    private static final BoundingBox2 TILE_BOUNDING_RECT = new BoundingBox2(0, 0, 1, 1);

    private final List<RectBody> rectList;

    public Player(){
        super(new BoundingBox2(0, 0, 2, 3));
        rectList = new ArrayList<>();
        getVelocity().setMax(50);
    }


    public void render(TextureBatch batch){
        batch.drawQuad(1, 1, 1, 1,  pos().x, pos().y, rect().getWidth(), rect().getHeight());
        
        for(RectBody r: rectList)
            batch.drawQuad(1, 0, 0, 0.5,  r.getMin().x, r.getMin().y, r.rect().getWidth(), r.rect().getHeight());
    }

    public void update(World world){
        WorldMap tileMap = world.getTileMap();

        // Getting the nearest tiles

        RectBody[] rects = getRects(tileMap, new Vec2f(), 1);

        final boolean isCollideUp = isCollide(0, Float.MIN_VALUE, rects);
        final boolean isCollideDown = isCollide(0, -Float.MIN_VALUE, rects);
        final boolean isCollideLeft = isCollide(-Float.MIN_VALUE, 0, rects);
        final boolean isCollideRight = isCollide(Float.MIN_VALUE, 0, rects);

        // Moving

        float delta = Pize.getDt();

        if(Key.A.isPressed())
            getVelocity().x -= 0.7;
        if(Key.D.isPressed())
            getVelocity().x += 0.7;

        // Auto jump

        if(isCollideDown && !isCollideUp){
            RectBody rectBody = this.copy();
            rectBody.pos().y++;

            if(
                (getVelocity().x > 0 && isCollideRight
                && !Collider2f.getCollidedMotion(rectBody, new Vec2f(Float.MIN_VALUE, 0), rects).isZero())
            ||
                (getVelocity().x < 0 && isCollideLeft
                && !Collider2f.getCollidedMotion(rectBody, new Vec2f(-Float.MIN_VALUE, 0), rects).isZero()
            ))
                getVelocity().y = 21;
        }

        // Gravity & Jump

        getVelocity().y -= 2;

        if(Key.SPACE.isPressed() && isCollideDown)
            getVelocity().y = 50;

        // Process collisions

        Vec2f motion = getVelocity().copy().mul(delta);
        rects = getRects(tileMap, motion, 0);
        Vec2f collidedVel = Collider2f.getCollidedMotion(this, motion, rects);

        getVelocity().reduce(0.5);
        getVelocity().collidedAxesToZero(collidedVel);
        getVelocity().clampToMax();

        pos().add(collidedVel);
    }

    public boolean isCollide(float x, float y, RectBody[] rects){
        return Collider2f.getCollidedMotion(this, new Vec2f(x, y), rects).isZero();
    }

    public RectBody[] getRects(WorldMap map, Vec2f vel, float padding){
        rectList.clear();
        for(int i = Maths.floor(getMin().x + vel.x - padding); i < getMax().x + vel.x + padding; i++)
            for(int j = Maths.floor(getMin().y + vel.y - padding); j < getMax().y + vel.y + padding; j++){

                MapTile tile = map.getTile(i, j);
                if(tile != null && tile.getType().collidable){
                    RectBody body = new RectBody(TILE_BOUNDING_RECT);
                    body.pos().set(i, j);
                    rectList.add(body);
                }
            }

        return rectList.toArray(new RectBody[0]);
    }

}

File: tests/test-terraria/src/main/kotlin/pize/tests/terraria/entity/Player.kt
package pize.tests.terraria.entity

import pize.Pize.dt
import pize.Pize.x
import pize.Pize.y
import pize.graphics.camera.Camera2D.x
import pize.graphics.camera.Camera2D.y
import pize.graphics.util.batch.TextureBatch
import pize.io.glfw.Key
import pize.math.Maths.floor
import pize.math.vecmath.vector.Vec2f
import pize.math.vecmath.vector.Vec2f.mul
import pize.physic.BoundingBox2
import pize.physic.Collider2f.getCollidedMotion
import pize.physic.RectBody
import pize.tests.terraria.map.WorldMap
import pize.tests.terraria.world.World

class Player : Entity(BoundingBox2(0.0, 0.0, 2.0, 3.0)) {
    private val rectList: MutableList<RectBody>

    init {
        rectList = ArrayList()
        velocity.setMax(50f)
    }

    override fun render(batch: TextureBatch) {
        batch.drawQuad(1.0, 1.0, 1.0, 1.0, pos().x, pos().y, rect()!!.width, rect()!!.height)
        for (r in rectList) batch.drawQuad(1.0, 0.0, 0.0, 0.5, r.min.x, r.min.y, r.rect()!!.width, r.rect()!!.height)
    }

    fun update(world: World?) {
        val tileMap = world.getTileMap()

        // Getting the nearest tiles
        var rects = getRects(tileMap, Vec2f(), 1f)
        val isCollideUp = isCollide(0f, Float.MIN_VALUE, rects)
        val isCollideDown = isCollide(0f, -Float.MIN_VALUE, rects)
        val isCollideLeft = isCollide(-Float.MIN_VALUE, 0f, rects)
        val isCollideRight = isCollide(Float.MIN_VALUE, 0f, rects)

        // Moving
        val delta = dt
        if (Key.A.isPressed) velocity.x -= 0.7.toFloat()
        if (Key.D.isPressed) velocity.x += 0.7.toFloat()

        // Auto jump
        if (isCollideDown && !isCollideUp) {
            val rectBody = copy()
            rectBody.pos().y++
            if ((velocity.x > 0 && isCollideRight
                        && !getCollidedMotion(rectBody, Vec2f(Float.MIN_VALUE, 0f), *rects).isZero)
                ||
                (velocity.x < 0 && isCollideLeft
                        && !getCollidedMotion(rectBody, Vec2f(-Float.MIN_VALUE, 0f), *rects).isZero)
            ) velocity.y = 21f
        }

        // Gravity & Jump
        velocity.y -= 2f
        if (Key.SPACE.isPressed && isCollideDown) velocity.y = 50f

        // Process collisions
        val motion: Vec2f = velocity.copy().mul(delta)
        rects = getRects(tileMap, motion, 0f)
        val collidedVel = getCollidedMotion(this, motion, *rects)
        velocity.reduce(0.5)
        velocity.collidedAxesToZero(collidedVel)
        velocity.clampToMax()
        pos().add(collidedVel)
    }

    fun isCollide(x: Float, y: Float, rects: Array<RectBody>): Boolean {
        return getCollidedMotion(this, Vec2f(x, y), *rects).isZero
    }

    fun getRects(map: WorldMap?, vel: Vec2f, padding: Float): Array<RectBody> {
        rectList.clear()
        var i = floor((min.x + vel.x - padding).toDouble())
        while (i < max.x + vel.x + padding) {
            var j = floor((min.y + vel.y - padding).toDouble())
            while (j < max.y + vel.y + padding) {
                val tile = map!!.getTile(i, j)
                if (tile != null && tile.type.collidable) {
                    val body = RectBody(TILE_BOUNDING_RECT)
                    body.pos()[i.toFloat()] = j.toFloat()
                    rectList.add(body)
                }
                j++
            }
            i++
        }
        return rectList.toTypedArray<RectBody>()
    }

    companion object {
        private val TILE_BOUNDING_RECT = BoundingBox2(0.0, 0.0, 1.0, 1.0)
    }
}
